users/tests/tests.py

```diff
@@ -34,9 +44,43 @@ def test_redirection_on_register_successful(self):
 
     def test_register_successful(self):
         response = self.client.post(reverse('users:register'), {'username': 'user_1', 'password1': 'pass1234', 'password2': 'pass1234'}, follow=True)
-        self.assertIn('You successfully created a new account!', response.content)
+        self.assertIn(UserRegistration.success_message, response.content)
 
     def test_register_unsuccessful(self):
         UserFactory(username='existing_user')
         response = self.client.post(reverse('users:register'), {'username': 'existing_user', 'password1': 'pass1234', 'password2': 'pass1234'})
         self.assertIn('A user with that username already exists.', response.content)
+
+
+class UserLoginIntegrationTests(TestCase, SimpleTestCase):
+    def test_redirection_on_login_successful(self):
+        UserFactory(username='existing_user')
+        response = self.client.post(reverse('users:login'), {'username': 'existing_user', 'password': 'pass1234'})
+        self.assertEqual(response.status_code, 302)
+
+    def test_login_successful(self):
+        UserFactory(username='existing_user')
+        response = self.client.post(reverse('users:login'), {'username': 'existing_user', 'password': 'pass1234'})
+        self.assertRedirects(response, reverse('users:user-profile'))
+
+    def test_login_unsuccessful(self):
+        response = self.client.post(reverse('users:login'), {'username': 'non_existing_user', 'password': 'pass1234'})
+        self.assertIn('Please enter a correct username and password. Note that both fields may be case-sensitive.', response.content)
+
+
+class UserProfileIntegrationTests(TestCase):
+    def test_user_profile_page_with_authenticated_user(self):
+        UserFactory(username='existing_user')
+        self.client.post(reverse('users:login'), {'username': 'existing_user', 'password': 'pass1234'})
+        response = self.client.get(reverse('users:user-profile'))
+        self.assertIn('You are logged in!', response.content)
+
+
+class AnonymousUserIntegrationTests(TestCase, SimpleTestCase):
+    def test_redirection_on_anonymous_user_access_profile_page(self):
+        response = self.client.get(reverse('users:user-profile'))
+        self.assertEqual(response.status_code, 302)
+
+    def test_redirection_to_login_on_anonymous_user_access_profile_page(self):
+        response = self.client.get(reverse('users:user-profile'))
+        self.assertRedirects(response, reverse('users:login') + '?next=' + reverse('users:user-profile'))
```

Comment:
```
self.assertRedirects(
    response, 
    '{redirect_path}?{next_path}'.format(
        redirect_path=reverse('users:login'),
        urllib.urlencode({'next': 'reverse('users:user-profile')'})
    ))
```

does this work? you should avoid string concat when possible. If you have to do it, you should DRY the logic of concatenating into a class or utility method (like urllib does for GET arguments). Also, format method is almost always the best solution for string assembly.